python/CommonFramework.cpp

```diff
@@ -63,93 +65,83 @@ bool PythonBase::Initialize()
-    if (!ExecScript(script)) {
-        ErrorLogger() << "Unable to redirect Python stdout and stderr";
-        return false;
-    }
+        // allow the "freeorion_logger" C++ module to be imported within Python code
+        try {
+            initfreeorion_logger();
+        } catch (...) {
+            ErrorLogger() << "Unable to initialize FreeOrion Python logging module";
+            return false;
+        }
 
-    // add the directory containing common Python modules used by all Python scripts to Python sys.path
-    if (!AddToSysPath(GetPythonCommonDir()))
-        return false;
+        // Allow C++ modules implemented by derived classes to be imported
+        // within Python code
 
-    // Allow C++ modules implemented by derived classes to be imported within Python code
-    if (!InitModules()) {
-        ErrorLogger() << "Unable to initialize FreeOrion Python modules";
+        if (!InitModules()) {
+            ErrorLogger() << "Unable to initialize FreeOrion Python modules";
+            return false;
+        }
+    } catch (error_already_set &err) {
+        HandleErrorAlreadySet();
         return false;
     }
 
     DebugLogger() << "FreeOrion Python interface successfully initialized!";
     return true;
 }
 
+bool PythonBase::IsPythonRunning()
+{ return Py_IsInitialized(); }
+
+void PythonBase::HandleErrorAlreadySet() {
+    if (!Py_IsInitialized()) {
+        ErrorLogger() << "Python interpreter not initialized and exception handler called.";
+        return;
+    }
+
+    // Matches system exit
+    if (PyErr_ExceptionMatches(systemExit.ptr()))
+    {
+        Finalize();
+        ErrorLogger() << "Python interpreter exited with SystemExit(), sys.exit(), exit, quit or some other alias.";
+        return;
+    }
+
+    PyErr_Print();
+    return;
+}
+
+
 void PythonBase::Finalize() {
-    if (m_python_interpreter_initialized) {
+    if (Py_IsInitialized()) {
         Py_Finalize();
-        m_python_interpreter_initialized = false;
+        DebugLogger() << "Cleaned up FreeOrion Python interface";
     }
-    DebugLogger() << "Cleaned up FreeOrion Python interface";
 }
 
-bool PythonBase::ExecScript(const std::string script) {
-    try {
-        object ignored = exec(script.c_str(), m_namespace, m_namespace);
-    }
-    catch (error_already_set err) {
-        PyErr_Print();
-        return false;
-    }
-    return true;
+void PythonBase::ExecScript(const std::string script) {
+    exec(script.c_str(), m_namespace, m_namespace);
 }
 
-bool PythonBase::SetCurrentDir(const std::string dir) {
+void PythonBase::SetCurrentDir(const std::string dir) {
     std::string script = "import os\n"
     "os.chdir(r'" + dir + "')\n"
     "print 'Python current directory set to', os.getcwd()";
-    if (!ExecScript(script)) {
-        ErrorLogger() << "Unable to set Python current directory";
-        return false;
-    }
-    return true;
+    exec(script.c_str(), m_namespace, m_namespace);
```

Comment:
Why don't use the `ExecScript` member function? That's the point of having it after all. If you use `exec` directly here and in `AddToSysPath`, it's actually obsolete and can be removed.